### PR TITLE
Ovs upstream for review

### PR DESCRIPTION
**Pull request by OphirMunk**

### CODE REVIEW

File: lib/netdev-rte-offloads.c [hunk 1]
```diff
@@ -731,3 +732,45 @@ netdev_rte_offloads_flow_del(struct netdev *netdev, const ovs_u128 *ufid,
...
+        NULLIFY(flow_flush);
+        NULLIFY(flow_dump_create);
+        NULLIFY(flow_dump_destroy);
+        NULLIFY(flow_dump_next);
+        NULLIFY(flow_put);
+        NULLIFY(flow_get);
+        NULLIFY(flow_del);
+        NULLIFY(init_flow_api);
+    }
+    netdev_vport_update_class(netdev, ndc);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: put inside the "if".
consider adding "error" message of "debug" in case of null

File: lib/netdev-vport.c [hunk 2]
```diff
@@ -150,6 +150,36 @@ netdev_vport_get_dpif_port(const struct netdev *netdev,
...
+        memcpy(vpc, ptr, sizeof *ptr);
+    }
+    ndc = &vpc->netdev_class;
+
+out:
+    return ndc;
+}
+
+void
+netdev_vport_update_class(struct netdev *netdev, struct netdev_class *ndc)
```
Comment (rbaryanai) [inline lib/netdev-vport.c]: why you need this function? you can do it on the rte-offload

File: lib/netdev-rte-offloads.c (hunk 1 above)
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: why not if ndc then netdev->netdev->class = ndc?

File: lib/netdev-vport.h [hunk 3]
```diff
@@ -42,6 +42,10 @@ void netdev_vport_inc_tx(const struct netdev *,
 bool netdev_vport_is_vport_class(const struct netdev_class *);
 const char *netdev_vport_class_get_dpif_port(const struct netdev_class *);
 
+struct netdev_class *netdev_vport_dup_class_once(struct netdev *netdev);
+void netdev_vport_update_class(struct netdev *netdev,
```
Comment (rbaryanai) [inline lib/netdev-vport.h]: don't think this is needed.

File: lib/netdev-vport.c (hunk 2 above)
Comment (OphirMunk) [reply lib/netdev-vport.c]: Reason: netdev_class struct is embedded inside vport_class, the structure of which is only known internally in netdev-vport.c.

File: lib/netdev-rte-offloads.c (hunk 1 above)
Comment (OphirMunk) [reply lib/netdev-rte-offloads.c]: Same reason as before: netdev_class struct is embedded inside vport_class, You must handle vport_class and not just netdev_class. Following your suggestions will result in a seg fault (based on my testing).

**Comment by yuskul** [comment]
bot:retest

**Comment by OphirMunk** [comment]
@rbaryanai @asafpenso please review

File: lib/netdev-rte-offloads.c [hunk 4]
```diff
@@ -765,16 +808,88 @@ netdev_rte_offloads_vxlan_init(struct netdev *netdev)
...
+        VLOG_DBG("port %d has no rte_port", dp_port);
+        return ENODEV;
     }
+
+    size_t hash = hash_bytes(&rte_port->port_no,
+                             sizeof rte_port->port_no, 0);
+    VLOG_DBG("Remove datapath port %d.", rte_port->port_no);
+    cmap_remove(&port_map, CONST_CAST(struct cmap_node *, &rte_port->node),
+                hash);
+   free(rte_port);
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: another space is needed at the beginning

File: lib/netdev-rte-offloads.c [hunk 5]
```diff
@@ -765,16 +808,88 @@ netdev_rte_offloads_vxlan_init(struct netdev *netdev)
...
+    cmap_insert(&port_map,
+                CONST_CAST(struct cmap_node *, &(*rte_port)->node), hash);
+next:
+    (*rte_port)->netdev = netdev;
+    (*rte_port)->port_no = dp_port;
+    (*rte_port)->rte_port_type = port_type;
+
+    return 0;
+}
+
```
Comment (OphirMunk) [inline lib/netdev-rte-offloads.c]: What is the returned value?

File: lib/netdev-rte-offloads.c [hunk 6]
```diff
@@ -28,8 +28,31 @@
...
+     RTE_PORT_TYPE_DPDK,
+     RTE_PORT_TYPE_VXLAN
+};
+
+/*
+ * A mapping from dp port number to port flow parameters.
+ */
+struct netdev_rte_port {
+    struct cmap_node node; /* Map by datapath port number. */
+    odp_port_t port_no; /* Datapath port number. */
```
Comment (OphirMunk) [inline lib/netdev-rte-offloads.c]: Consider renaming port_no ==> odp_port

File: lib/netdev-rte-offloads.c [hunk 7]
```diff
@@ -727,12 +750,32 @@ netdev_rte_offloads_flow_del(struct netdev *netdev, const ovs_u128 *ufid,
...
+static struct netdev_rte_port *
+netdev_rte_port_search(odp_port_t port_no, struct cmap *map)
+{
+    size_t hash = hash_bytes(&port_no, sizeof port_no, 0);
+    struct netdev_rte_port *data;
+
+    VLOG_DBG("Search for port %d.", port_no);
+
+    CMAP_FOR_EACH_WITH_HASH (data, node, hash, map) {
+        if (port_no ==  data->port_no) {
```
Comment (OphirMunk) [inline lib/netdev-rte-offloads.c]: Remove double spaces == ...

File: lib/netdev-rte-offloads.c [hunk 8]
```diff
@@ -727,12 +750,32 @@ netdev_rte_offloads_flow_del(struct netdev *netdev, const ovs_u128 *ufid,
...
+/*
+ * Search for offloaded port by odp port no.
+ */
+static struct netdev_rte_port *
+netdev_rte_port_search(odp_port_t port_no, struct cmap *map)
+{
+    size_t hash = hash_bytes(&port_no, sizeof port_no, 0);
+    struct netdev_rte_port *data;
+
+    VLOG_DBG("Search for port %d.", port_no);
```
Comment (OphirMunk) [inline lib/netdev-rte-offloads.c]: Remove DBG message which will overflow the log

File: lib/netdev-rte-offloads.c (hunk 4 above)
Comment (OphirMunk) [reply lib/netdev-rte-offloads.c]: Fixed

**Pull request closed without merging**